_drafts/unit-test-mocking.md

```diff
@@ -0,0 +1,1006 @@
+  CHECK_TRUE(success);
+
+  // Push the raw bytes into our protocol
+  size_t len = s_resp_buffer_len;
+  eProtocolCode rv = protocol_handle(
+      in_bytes, sizeof(in_bytes),
+      s_resp_buffer, &len);
+
+  // Ensure we receive the correct response!
+  CHECK_EQUAL(kProtocolCode_Ok, rv);
+  MEMCMP_EQUAL(val_bytes, s_resp_buffer, sizeof(val_bytes));
+}
+```
+
+We are passing in a set of raw bytes into our protocol handler, and essentially
+letting it run its course. This data flows in and out of **five** layers of
+code, as shown in our diagram, and eventually returns data to our unit test with
+`s_resp_buffer`. We confirm the data is correct using CppUTest's `MEMCMP_EQUAL`
+operation.
+
+Feel free to check out the
+[source code on Github](https://github.com/memfault/interrupt/example/unit-testing/complex/tests/src/test_integration_protocol_kv_store.cpp)
+
+> Do these integration tests give us the optimal 100% real-world test that QA or
+> in-field testing provides? Of course not, but it gets us 90% of the way there
+> which is infinitely better than nothing.
+
+### Integration Tests in Practice
+
+This type of integration test can us embedded software developers a lot of bang
+for our buck, but also cause frustration.
+
+#### Benefits
+
+- A large amount of code and logic can be tested using [continuous
+  integration]({% post_url 2019-09-17-continuous-integration-for-firmware %})
+- Ensuring correctness after a large refactor of a module that resides in the
+  middle of an integration test becomes trivial.
+- Enables short-staffed and budget-conscious teams to push out building a
+  complex test automation system with real hardware using something like Robot
+  Framework[^robot_framework].
+
+#### Drawbacks
+
+- Changing the API of a single commonly used module within integration tests
+  will require updating all of those tests.
+- They are susceptible to breaking in undefined ways since many source files are
+  thrown into the same test.
+- Integration tests are, in some ways, easier to write since they don't require
+  new stub, fake, or mock files for each test that gets written. They might
+  become the new normal, which would mean that each module would become less
+  well tested and coverage of 100% might be more difficult to attain.
+
+With the pros and cons taken into account, I can't argue with any amount of
+testing, especially if it's fast, reliable, and tests the code in meaningful
+ways. Even if I've battled a stale unit or integration test 2 years after it was
+written, I reminded myself that I **would not** want to be testing it manually.
+
+## Final Thoughts
+
+I'm a believer that proper unit testing and integration testing can prevent many
+of the firmware bugs that devices experience in the field. Except for the
+firmware code that directly pokes in register values into hardware, all of the
+software that goes into embedded software can and should be unit testing in some
+fashion.
```

Comment:
Maybe: 

Could add some verbiage about others sharing how they've used unit tests and the "See anything you’d like to change? Submit a pull request or open an issue at Github" if you want